fix: Fix invalid ngettext usage

last-partizan opened the pull request:
## Description

Format should be called after `ngettext`, not before.

See example: https://docs.djangoproject.com/en/5.1/topics/i18n/translation/#pluralization

Formatting changes is because i don't know how to properly format it with your previous style, and there's no included autoformat/or formatting linter. I just formatted it automatically with ruff-lsp. Feel free to revert fomatting changes to your style or explain me what to do.

auvipy reviewed (changes requested):
can you please share the output of before and after this change?

last-partizan commented:
@auvipy please take a look at this again.

auvipy reviewed (changes requested):
is their anyway you can ensure that this won't create any regression, please?

last-partizan commented:
![image](https://github.com/user-attachments/assets/be77fe29-60a3-42ce-9382-70fa290e73b3)

This code is covered by tests, and translation strings is unchanged, so everything should be fine.

We probably need to make sure changelog is clear about what's changed here. I have squashed commits and updated text, but if changelog is autogenerated from the first line of commit, you may need to move relevant instructions into the first line when Merging/Squashing.

Important part there: "If you have overridden `extra_detail_singular` or `extra_detail_plural`,  you should now replace them with a single `def extra_detail` override.".

auvipy commented:
we will need documentation update then? and it seems to be an API change as well?

last-partizan commented:
> This is an API change, which the project prefers not to do at this stage.

By "at this stage", you mean "project is mature enough to not do API changes"? or some sort of "no API changes for minor releases"?

> It may be possible to juggle with some properties etc. to strictly extend the previous API without removing `Throttled.extra_detail_*` attributes, but I'm not sure that complexity would be a good idea.

Not sure how. Even if we can use something like `ngettext(self.extra_detail_plural ...`, it will not work, because for ngettext to work properly it needs strings right there.

And, I don't see any docs mentioning `extra_detail_` on class Throttled, so maybe this is not a public API...

peterthomassen commented:
> By "at this stage", you mean "project is mature enough to not do API changes"? or some sort of "no API changes for minor releases"?

The former, see https://github.com/encode/django-rest-framework/blob/master/CONTRIBUTING.md and the note on https://www.django-rest-framework.org/community/contributing/.

> And, I don't see any docs mentioning `extra_detail_` on class Throttled, so maybe this is not a public API...

Many people consider all API public unless identifiers start with an underscore.

peterthomassen commented:
@tomchristie Thoughts?

peterthomassen commented on rest_framework/exceptions.py:
```diff
@@ -226,23 +226,25 @@ def __init__(self, media_type, detail=None, code=None):
         self.wait = wait
         super().__init__(detail, code)
 
+    @staticmethod
+    def extra_detail(wait):
+        return ngettext(
+            'Expected available in {wait} second.',
+            'Expected available in {wait} seconds.',
+            wait,
+        ).format(wait=wait)
```
Suggestion:
- Set `extra_detail_singular` and `extra_detail_plural` as they used to be, but with out the `_()` call (so, just the English strings)
- Make `extra_detail` a class method as follows:
  ```python
    @classmethod
    def extra_detail(cls, wait):
        return ngettext(
            cls.extra_detail_singular,
            cls.extra_detail_plural,
            wait,
        ).format(wait=wait)
  ```
- Keep the change in `__init__`

This way, code using the old attributes will not start failing.

auvipy replied:
Great suggestion. If this PR helps to improve django compatibility, then we might consider it

last-partizan replied:
@peterthomassen we cannot use plain strings here, because `django-admin makemessages` needs them to be inside `ngettext` function call, to properly create msgid.

Here's changes to `django.po` file after running `django-admin makemessages -l en_US`

Note that we now have instead of two independently translated strings single "msgid" + "msgid_plural".

If we use plain strings, they just disappear from translation file.

```diff
-#: exceptions.py:224
+#: exceptions.py:243
 #, python-brace-format
 msgid "Expected available in {wait} second."
-msgstr ""
+msgid_plural "Expected available in {wait} seconds."
+msgstr[0] ""
+msgstr[1] ""
 
-#: exceptions.py:225
-#, python-brace-format
-msgid "Expected available in {wait} seconds."
-msgstr ""
-
```

Maybe I should commit this file to make changes noticeable, but I was thinking it's supposed to be updated during maintenance/translation stage. And if I commit it now, it will have some unrelated changes as well.

peterthomassen replied:
> needs them to be inside `ngettext` function call

They are, please see https://github.com/encode/django-rest-framework/pull/9560#discussion_r1949163024

last-partizan replied:
Let me check if I understood you correctly.

@peterthomassen you are suggesting doing this, right?

```diff
diff --git a/rest_framework/exceptions.py b/rest_framework/exceptions.py
index 09f11110..7cfa5dfb 100644
--- a/rest_framework/exceptions.py
+++ b/rest_framework/exceptions.py
@@ -226,8 +226,8 @@ class UnsupportedMediaType(APIException):
 class Throttled(APIException):
     status_code = status.HTTP_429_TOO_MANY_REQUESTS
     default_detail = _('Request was throttled.')
-    extra_detail_singular = _('Expected available in {wait} second.')
-    extra_detail_plural = _('Expected available in {wait} seconds.')
+    extra_detail_singular = 'Expected available in {wait} second.'
+    extra_detail_plural = 'Expected available in {wait} seconds.'
     default_code = 'throttled'
 
     def __init__(self, wait=None, detail=None, code=None):
@@ -237,9 +237,9 @@ class Throttled(APIException):
             wait = math.ceil(wait)
             detail = ' '.join((
                 detail,
-                force_str(ngettext(self.extra_detail_singular.format(wait=wait),
-                                   self.extra_detail_plural.format(wait=wait),
-                                   wait))))
+                force_str(ngettext(self.extra_detail_singular,
+                                   self.extra_detail_plural,
+                                   wait).format(wait=wait))))
         self.wait = wait
         super().__init__(detail, code)
 ```

This will lead to removal of strings from translation files.

```diff
-#: exceptions.py:224
-#, python-brace-format
-msgid "Expected available in {wait} second."
-msgstr ""
-
-#: exceptions.py:225
-#, python-brace-format
-msgid "Expected available in {wait} seconds."
-msgstr ""
-
-#: fields.py:316 relations.py:245 relations.py:279 validators.py:90
-#: validators.py:183
+#: fields.py:292 relations.py:240 relations.py:276 validators.py:99
+#: validators.py:219
 msgid "This field is required."
 msgstr ""
```

Because makemessages does not evaluate python code, only parses it. So we need string literals to be present in function calls here.

peterthomassen replied:
> Because makemessages does not evaluate python code, only parses it. So we need string literals to be present in function calls here.

Oh, I see. That seems very unfortunate.

last-partizan replied:
@auvipy this PR improves Django compatibility, because it replaces incorrect usage of ngettext with correct one :)

If we want to be completely backward-compatible here, maybe use something like 

```
if hasattr(self, 'extra_detail_singular'): ... use old method
else: ... use proper method
```

But I don't think it's necessary to support wrong gettext usage, even it was once used.

peterthomassen replied:
I think adding this makes sense.

peterthomassen commented:
@tomchristie wdyt?

tomchristie commented:
Not really worth our time.

Least disruptive change would be...

```python
        return ngettext(
            self.extra_detail_singular,
            self.extra_detail_plural,
            wait,
        ).format(wait=wait)
```

but doesn't meet our guidelines for winding down churn. (shrug)

last-partizan commented:
Current variant is already 0-disruptive. It fixed bug, allows redefinition, and warns user when they're using incorrect variant.

And what @tomchristie suggests does not work

>  Because makemessages does not evaluate python code, only parses it. So we need string literals to be present in function calls here.

https://github.com/encode/django-rest-framework/pull/9560#discussion_r1950702935

tomchristie commented:
We can see the behavior change. The cost is not worthwhile in my judgement.

More to the point I think what @tomchristie is *actually* suggesting is that they're no longer willing to tolerate *either intentional or unintentional* social DDOS'ing from internet anons. Somehow this became a dominant theme within OSS development. As it stands REST framework will not be accepting further contributions.
